pkg/informer/informer.go

```diff
@@ -787,6 +847,114 @@ func newRESTMapper(fn func() (meta.RESTMapper, error)) restMapper {
 	}
 }
 
+func NewCRDGVRSource(informer cache.SharedIndexInformer) (*crdGVRSource, error) {
+	// Add an index function that indexes a CRD by its group/version/resource.
+	if err := informer.AddIndexers(cache.Indexers{
+		byGroupVersionResourceIndex: byGroupVersionResourceIndexFunc,
+	}); err != nil {
+		return nil, err
+	}
+
+	return &crdGVRSource{
+		crdInformer: informer,
+		crdIndexer:  informer.GetIndexer(),
+	}, nil
+
+}
+
+func byGroupVersionResourceKeyFunc(group, version, resource string) string {
+	return fmt.Sprintf("%s/%s/%s", group, version, resource)
+}
+
+func byGroupVersionResourceIndexFunc(obj interface{}) ([]string, error) {
+	crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+	if !ok {
+		return nil, fmt.Errorf("%T is not a CustomResourceDefinition", obj)
+	}
+
+	var ret []string
+
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		ret = append(ret, byGroupVersionResourceKeyFunc(crd.Spec.Group, v.Name, crd.Spec.Names.Plural))
+	}
+
+	return ret, nil
+}
+
+type crdGVRSource struct {
+	crdInformer cache.SharedIndexInformer
+	crdIndexer  cache.Indexer
+}
+
+func (s *crdGVRSource) GVRs() map[schema.GroupVersionResource]GVRPartialMetadata {
+	crdGVRs := s.crdIndexer.ListIndexFuncValues(byGroupVersionResourceIndex)
+	result := make(map[schema.GroupVersionResource]GVRPartialMetadata, len(crdGVRs))
+	for _, indexedValue := range crdGVRs {
+		parts := strings.Split(indexedValue, "/")
+		gvr := gvrFor(parts[0], parts[1], parts[2])
+
+		obj, err := indexers.ByIndex[*apiextensionsv1.CustomResourceDefinition](s.crdIndexer, byGroupVersionResourceIndex, byGroupVersionResourceKeyFunc(gvr.Group, gvr.Version, gvr.Resource))
+		if err != nil {
+			utilruntime.HandleError(err)
+			continue
+		}
+		if len(obj) == 0 {
+			utilruntime.HandleError(fmt.Errorf("unable to retrieve CRD for GVR: %s", gvr))
+			continue
+		}
+		// We assume CRDs partial metadata for the same GVR are constant
+		crd := obj[0]
+		result[gvr] = withGVRPartialMetadata(crd.Spec.Scope, crd.Spec.Names.Kind, crd.Spec.Names.Singular)
+	}
+	return result
+}
+
+func (s *crdGVRSource) Synced() bool {
+	return s.crdInformer.HasSynced()
+}
+
+func (s *crdGVRSource) crdIsEstablished(obj interface{}) bool {
+	crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unable to determine if CRD is established - got type %T instead", obj))
+		return false
+	}
+
+	return apihelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established)
+}
+
+func (s *crdGVRSource) Changes() <-chan struct{} {
+	changes := make(chan struct{})
```

Comment:
We might need to make this buffered (1) and then use a `select` statement to try to send to the channel, or drop if already full. Otherwise, the event handler functions below block until a receiver can pop from the channel.

Reply:
The channel returned by `Changes()` is itself "forwarded" to a channel which follows the rules you're mentioning: https://github.com/davidfestal/kcp/blob/b5eef4f55befb1492d2171c9b3b320a22c3cf016/pkg/informer/informer.go#L179
Isn't it sufficient ?

Reply:
Fixed in Commit d991a46252745e8c5dffe3c7df404c214559bcac